poly-commitment/src/commitment.rs

```diff
@@ -607,22 +607,22 @@ impl<G: CommitmentCurve> SRS<G> {
         &self,
         domain: D<G::ScalarField>,
         plnm: &Evaluations<G::ScalarField, D<G::ScalarField>>,
-        max: Option<usize>,
```

Comment:
at this point I'm not sure why this is removed. I know we don't use it but it seems like it's a worthy API to keep around

Reply:
This is for evaluations. The coefficients version is still around, but the evaluations form never made any sense.

In my experience, reviewing commit by commit is nearly always unwise, unless the author explicitly recommends it. The whole point of this PR is the [final commit](https://github.com/o1-labs/proof-systems/pull/870/commits/9c82400b649340bbdc9c25deb2148a94062e15ba), things were only split this way so that I could test as I went.

Reply:
why wouldn't you want to enforce an upperbound on a polynomial because it's in evaluation form? Not sure I get it

Reply:
also now at this point I'm wondering why moving from `commit_helper` to `multi_scalar_mul` which doesn't do any chunking. What if we're trying to commit to a polynomial larger than the SRS?

Reply:
The `multi_scalar_mul` is against the lagrange basis, which is now a `PolyComm`. That is to say: the lagrange basis encodes chunked commitments of size `domain.d1` and chunk size of `srs.len()`.

It's also worth mentioning: the old code was always broken. Attempting to 'chunk' some evaluations would *chunk the evaluations*, rather than the underlying polynomial, which is never what you want. Consider modifying an evaluation by `x` at the index `i`. The contribution of this evaluation to the entire polynomial is `x * k_i * (x^n - 1) / (x - w^i)`, where `k_i` is the Lagrange polynomial normalisation factor.

So, when chunking correctly, every evaluation will have an effect on every chunk, due to the `x^(n-1)` terms in that expression above. However, chunking the evaluations results in each evaluation only affecting the chunk matching its index, resulting in a junk commitment.

Reply:
In case it's helpful, lets work through an example with a domain size of 4 and a srs size of 2. I'll use complex numbers (1, i, -1, -i) so that I don't need to remember what sqrt(-1) is in the pasta curves.

Suppose we have the degree 4 polynomial
```
f(x) = 1 + 2 x + 3 x^2 + 4 x^3
```
and thus evaluations
```
[f(1), f(i), f(-1), f(-i)]
```
which we can compute directly
```
[10, -2 - 2i, -2, -2 + 2i]
```

The chunked version of `f` has 2 parts:
```
f_low(x) = 1 + 2 x
f_high(x) = 3 + 4 x
```

We can also compute the Lagrange normalisation factors for each term, those being `1/l_i(w^i)` where `l_i(x) = (x^4 - 1)/(x - w^i)`. Recall that `x^4 - 1 = (x - w)(x - w^2)(x - w^3)(x - w^4)`.

```
l_0(1) = (1 - i)(1 + i)(1 + 1) = 2(2 + 1) = 6
l_1(i) = (i - 1)(i + i)(i + 1) = 2i(-1 - 1) = -4i
l_2(-1) = (-1 - 1)(-1 - i)(-1 + i) = -2(1 + 1) = -4
l_3(-i) = (-i - 1)(-i - i)(-i + 1) = -2i(-1 - 1) = 4i
```

Correspondingly, if we change the first evaluation, e.g. by increasing it by 6, we modify the original polynomial by `6/l_0(w^0) * (x^4 - 1)/(x - w^0)`, which expands to
```
  1/1 * (x^4                                           - 1)     / (x - 1)
=       (x^4 - x^3   + x^3 - x^2   + x^2 - x^1   + x^1 - x^0)   / (x - 1)
=       (x^3 (x - 1) + x^2 (x - 1) + x^1 (x - 1) + x^0 (x - 1)) / (x - 1)
=       (x^3         + x^2         + x^1         + x^0)
```

We can check this by adding this into the original polynomial:
```
f(x) = 1 + 2 x + 3 x^2 + 4 x^3
g(x) = 2 + 3 x + 4 x^2 + 5 x^3
```
and evaluating `g` at the domain
```
[g(1), g(i), g(-1), g(-i)] = [16, -2 - 2i, -2, -2 + 2i]
```

------------------------

Chunking `g(x)` gives
```
g_low(x) = 2 + 3 x
g_high(x) = 4 + 5 x
```

So we have that
* the lagrange polynomial `L_1(x) = 1/6 (x^3 + x^2 + x^1 + x^0)`
* the chunked lagrange polynomials are `L_1_low(x) = 1/6 (1 + x)`, `L_1_high(x) = 1/6 (1 + x)`.

The same applies for each of the other evaluations: we need to consider their impact *on each chunk of the polynomial* in order to capture their effect on the entire polynomial faithfully. We then use the ifft trick to compute these contributions efficiently as commitments.

Since commitments are additively homomorphic (equiv. a vector field representing the polynomials), we can do a multi-scalar-multiplication to generate a commitment to the entire polynomial from the evaluations and the commitments to the corresponding chunked lagrange polynomials.